core/include/gnuradio-4.0/Block.hpp

```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+                    if (stride.value != 0 && stride.value != denominator) {
+                        nResamplingChunks = denominator <= maxSyncIn && numerator <= maxSyncOut ? 1 : 0;
+                    } else {
+                        nResamplingChunks = std::min(maxSyncIn / denominator, maxSyncOut / numerator);
+                    }
+                } else {
+                    nResamplingChunks = std::min(maxSyncIn / denominator, maxSyncOut / numerator);
+                }
+                if (nResamplingChunks * denominator < minSyncIn || nResamplingChunks * numerator < minSyncOut) {
+                    return std::tuple{0UZ, 0UZ};
+                }
+                return std::tuple{static_cast<std::size_t>(nResamplingChunks * denominator), static_cast<std::size_t>(nResamplingChunks * numerator)};
+            } else {
+                std::size_t n = std::min(maxSyncIn, maxSyncOut);
+                return std::tuple{n, n};
+            }
+        } else { // no resampling
+            std::size_t n = std::min(maxSyncIn, maxSyncOut);
+            return std::tuple{n, n};
+        }
+    }
+
+    auto
+    getMergedBlockLimit() {
+        if constexpr (requires(const Derived &d) {
+                          { available_samples(d) } -> std::same_as<std::size_t>;
+                      }) {
+            return available_samples(self());
+        } else if constexpr (traits::block::stream_input_port_types<Derived>::size == 0
+                             && traits::block::stream_output_port_types<Derived>::size
+                                        == 0) { // allow blocks that have neither input nor output ports (by merging source to sink block) -> use internal buffer size
+            constexpr gr::Size_t chunkSize = Derived::merged_work_chunk_size();
+            static_assert(chunkSize != std::dynamic_extent && chunkSize > 0, "At least one internal port must define a maximum number of samples or the non-member/hidden "
+                                                                             "friend function `available_samples(const BlockType&)` must be defined.");
+            return chunkSize;
+        }
+        return std::numeric_limits<std::size_t>::max();
+    }
+
+    work::Status
+    invokeProcessBulk(auto &inputSpans, auto &writersTuple) {
+        // cannot use std::apply because it requires tuple_cat(inputSpans, writersTuple). The latter doesn't work because writersTuple isn't copyable.
+        return [&]<std::size_t... InIdx, std::size_t... OutIdx>(std::index_sequence<InIdx...>, std::index_sequence<OutIdx...>) {
+            return self().processBulk(std::get<InIdx>(inputSpans)..., std::get<OutIdx>(writersTuple)...);
+        }(std::make_index_sequence<std::tuple_size_v<std::remove_cvref_t<decltype(inputSpans)>>>(), std::make_index_sequence<std::tuple_size_v<std::remove_cvref_t<decltype(writersTuple)>>>());
+    }
```

Comment:
Not (yet) fully tested but the below could convert the present `std::vector<std::span>` or `std::vector<"ConsumableSpan">` to `std::span<std::span..> or `std::span<"ConsumableSpan">`

```cpp
work::Status invokeProcessBulk(auto &inputSpans, auto &writersTuple) {
    auto convertIfNeeded = [](auto&& arg) -> decltype(auto) {
        using T = std::decay_t<decltype(arg)>;
        if constexpr (std::is_same_v<T, std::vector<typename T::value_type>>>) {
            // convert std::vector to std::span of [Consumable,Producable]Spans
            return std::span<typename T::value_type>{arg.data(), arg.size()};
        } else if (...) {  // similar for std::array<...>
        } else { // pass-through for other types
            return std::forward<decltype(arg)>(arg);
        }
    };

    return [&]<std::size_t... InIdx, std::size_t... OutIdx>(std::index_sequence<InIdx...>, std::index_sequence<OutIdx...>) {
        return self().processBulk(convertIfNeeded(std::get<InIdx>(inputSpans))..., convertIfNeeded(std::get<OutIdx>(writersTuple))...);
    }(std::make_index_sequence<std::tuple_size_v<std::remove_cvref_t<decltype(inputSpans)>>>(), std::make_index_sequence<std::tuple_size_v<std::remove_cvref_t<decltype(writersTuple)>>>());
}
```

@drslebedev your view?

Reply:
N.B. this way we do not expose if the dynamic input ports/spans are stored as vector or array and makes IMO a more maintainable and future-proof  API. Feedback is welcome.

Reply:
Using `std::span` for an array of ports as the signature for processBulk() looks good to me. 
`std::span<std::span..>` is better than `std::vector<std::span..>`.

However, I recommend postponing this to the next pull request, as it requires further modifications in the unit tests.